### ISSUE TITLE
Passing fs_storage_options nested dict with client_kwargs to ParquetDataCatalog throws `TypeError`

### ISSUE THREAD

**Issue by dimitar-petrov** [nautechsystems/nautilus_trader#1504]
# Bug Report

### Expected Behavior
To create datacatalog with custom s3 endpoint by passing storage_options in the following format:
```py
{'client_kwargs': {'endpoint_url': 'https://minio.example.com'},
 'key': 'key',
 'secret': 'secret',
 'use_ssl': True}
```

### Actual Behavior
There is issue inside the singleton with passing a nested dict as fs_storage_options but it is a valid fsspec parameter.
```py
ipdb> n
TypeError: unhashable type: 'dict'
> /home/user/.pex/installed_wheels/a5cd33a35d3178439871cb01ebd6cd111bb3d64c684ff6c275fb999e4adc8f7a/nautilus_trader-1.186.0-cp310-cp310-manylinux_2_39_x86_64.whl/nautilus_trader/persistence/catalog/singleton.py(37)__call__()
     32         full_kwargs = resolve_kwargs(cls.__init__, None, *args, **kw)
     33         if full_kwargs == {"self": None, "args": (), "kwargs": {}}:
     34             full_kwargs = {}
     35         full_kwargs.pop("self", None)
     36         key = tuple(full_kwargs.items())
---> 37         if key not in cls._instances:
     38             cls._instances[key] = super().__call__(*args, **kw)
     39         return cls._instances[key]
     40 
     41 

ipdb> cls._instances
{}
ipdb> key
(('path', 'test-bucket/newcatalog'), ('fs_protocol', 's3'), ('fs_storage_options', (('client_kwargs', {'endpoint_url': 'https://minio.example.com'}), ('key', 'key'), ('secret', 'secret'), ('use_ssl', True))), ('dataset_kwargs', None), ('min_rows_per_group', 0), ('max_rows_per_group', 5000), ('show_query_paths', False))
```

### Steps to Reproduce the Problem
1. Create datacatalog as follows
```py
catalog = ParquetDataCatalog(
    "test-bucket/newcatalog",
    fs_protocol="s3",
    fs_storage_options={
        'client_kwargs': {'endpoint_url': 'https://minio.example.com'},
        'key': 'key',
        'secret': 'secret',
        'use_ssl': True}
)
```

### Specifications

- OS platform: archlinux
- Python version: 3.10
- `nautilus_trader` version: 1.186.0

**Comment by limx0** [comment]
Hi @dimitar-petrov,

Thanks for the reproducible example.

 You can work around this limitation on the ParquetDataCatalog by simply wrapping your client_kwargs in a FrozenDict like: 

```python
from frozendict import frozendict

catalog = ParquetDataCatalog(
    "test-bucket/newcatalog",
    fs_protocol="s3",
    fs_storage_options={
        'client_kwargs': frozendict({'endpoint_url': 'https://minio.example.com'}),
        'key': 'key',
        'secret': 'secret',
        'use_ssl': True}
)
```

We may improve this in the future but hopefully this workaround is acceptable for now.

**Comment by dimitar-petrov** [comment]
@limx0, Thanks, a nice workarround.

I took the following route to patch `freeze_dict` func.

```py
modified   nautilus_trader/persistence/catalog/singleton.py
@@ -56,4 +56,4 @@ def check_value(v: Any) -> Any:
 
 
 def freeze_dict(dict_like: dict) -> tuple:
-    return tuple(sorted(dict_like.items()))
+    return tuple(sorted((k, check_value(v)) for k, v in dict_like.items()))
```

Let me know whether to open PR.

And a qick question on the side.
Is it the case that it s3 is not implemented on the rust side yet.
Noticed that reading from s3 is not possible so it diminishes the use case.

**Comment by limx0** [comment]
Hi @dimitar-petrov,

> Let me know whether to open PR.

Yes we would welcome any PRs here. 

> Is it the case that it s3 is not implemented on the rust side yet.

I'm not sure exactly what you mean here - reading from s3 is definitely supported (but you will need to install s3fs in your environment)

**Comment by limx0** [comment]
Ah yes, apologies. For rust queries we have not yet implemented non filesystem backends. PRs welcome.

**Comment by cjdsellers** [comment]
Hi @dimitar-petrov

I've fixed this on `develop` branch now, thanks for providing the diff there (I've also credited you for the fix in the release notes).

I'll close this ticket for now as it seems we're at an enhancement request to support non-filesystem backends for Rust datafusion.
Feel free to raise another specific enhancement request ticket for this, including your suggestions/requirements where we could discuss further. 

As @limx0 says we certainly welcome any PRs which could help towards this feature.

Many thanks!
